pkg/controllers/v1alpha1/datamigrate/implement.go

```diff
@@ -0,0 +1,383 @@
+			DataMigrate, err := utils.GetDataMigrate(r.Client, targetDataMigrate.Name, targetDataMigrate.Namespace)
+			if err != nil {
+				return err
+			}
+			DataMigrateToUpdate := DataMigrate.DeepCopy()
+			DataMigrateToUpdate.Status.Phase = common.PhaseFailed
+			DataMigrateToUpdate.Status.Infos = map[string]string{}
+
+			if !reflect.DeepEqual(DataMigrateToUpdate.Status, DataMigrate.Status) {
+				if err := r.Status().Update(ctx, DataMigrateToUpdate); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			log.Error(err, "can't update DataMigrate's phase status to Failed")
+			return utils.RequeueIfError(err)
+		}
+		return utils.NoRequeue()
+	}
+
+	// 2. Check if there's any Executing DataMigrate jobs(conflict DataMigrate)
+	conflictDataMigrateRef := targetDataSet.Status.DataMigrateRef
+	myDataMigrateRef := utils.GetDataMigrateRef(targetDataMigrate.Name, targetDataMigrate.Namespace)
+	if len(conflictDataMigrateRef) != 0 && conflictDataMigrateRef != myDataMigrateRef {
+		log.V(1).Info("Found other DataMigrates that is in Executing phase, will backoff", "other DataMigrate", conflictDataMigrateRef)
+		r.Recorder.Eventf(&targetDataMigrate,
+			v1.EventTypeNormal,
+			common.DataMigrateCollision,
+			"Found other DataMigrate(%s) that is in Executing phase, will backoff",
+			conflictDataMigrateRef)
+		return utils.RequeueAfterInterval(20 * time.Second)
+	}
+
+	// 3. Check if the bounded runtime is ready
+	ready := engine.CheckRuntimeReady()
+	if !ready {
+		log.V(1).Info("Bounded accelerate runtime not ready", "targetDataset", targetDataSet)
+		r.Recorder.Eventf(&targetDataMigrate,
+			v1.EventTypeNormal,
+			common.RuntimeNotReady,
+			"Bounded accelerate runtime not ready")
+		return utils.RequeueAfterInterval(20 * time.Second)
+	}
+
+	// 4. lock the target dataset. Make sure only one DataMigrate can win the lock and
+	// the losers have to requeue and go through the whole reconciliation loop.
+	log.Info("No conflicts detected, try to lock the target dataset")
+	datasetToUpdate := targetDataSet.DeepCopy()
+	datasetToUpdate.Status.DataMigrateRef = myDataMigrateRef
+	if !reflect.DeepEqual(targetDataSet.Status, datasetToUpdate.Status) {
+		if err := r.Client.Status().Update(context.TODO(), datasetToUpdate); err != nil {
+			log.V(1).Info("fail to get target dataset's lock, will requeue")
+			//todo(xuzhihao): random backoff
+			return utils.RequeueAfterInterval(20 * time.Second)
+		}
+	}
+
+	// 5. update dataset phase to DataMigrating
+	log.Info("Get lock on target dataset, try to update phase")
+	DataSetToUpdate := targetDataSet.DeepCopy()
+	DataSetToUpdate.Status.Phase = datav1alpha1.DataMigrating
+	if err := r.Client.Status().Update(context.TODO(), DataSetToUpdate); err != nil {
+		log.Error(err, "failed to update DataSet's status to DataMigrating, will retry")
+		return utils.RequeueIfError(err)
+	}
+	log.V(1).Info("update DataSet's status to DataMigrating successfully")
```

Comment:
Suggest to combine step 4 and step 5 to make less update operations to api server.